src/coreclr/src/tools/Common/TypeSystem/Common/Utilities/LockFreeReaderHashtable.cs

```diff
@@ -227,16 +288,32 @@ private void Expand(TValue[] oldHashtable)
                 // Work in a local variable to avoid lots of unnecessary volatile reads of _newHashTable since only this method can
                 // change it and we're under a lock
                 TValue[] newHashTable = new TValue[newSize];
-                _newHashTable = newHashTable;
+                Interlocked.Exchange(ref _newHashTable, newHashTable);
```

Comment:
Perhaps worth adding a comment so this is not "simplified" into  `_newHashTable = newHashTable` later. 
This is a rare  "read-after-write" case were even x64/x86 need fences.

Basically - we must be sure that the publishing of `_newHashTable` happens before we read the first table entry from the point of view of external observer. 
Without full fence the write buffer may delay the write until after we start reading when it will be too late.